modules/core/02-client/keeper/client.go

```diff
@@ -9,192 +9,191 @@ import (
+func (k Keeper) CreateClient(ctx sdk.Context, clientType string, clientState, consensusState []byte) (string, error) {
+	if clientType == exported.Localhost {
+		return "", errorsmod.Wrapf(types.ErrInvalidClientType, "cannot create client of type: %s", clientType)
 	}
 
 	params := k.GetParams(ctx)
-	if !params.IsAllowedClient(clientState.ClientType()) {
+	if !params.IsAllowedClient(clientType) {
 		return "", errorsmod.Wrapf(
 			types.ErrInvalidClientType,
-			"client state type %s is not registered in the allowlist", clientState.ClientType(),
+			"client state type %s is not registered in the allowlist", clientType,
 		)
 	}
 
-	clientID := k.GenerateClientIdentifier(ctx, clientState.ClientType())
-	clientStore := k.ClientStore(ctx, clientID)
+	clientID := k.GenerateClientIdentifier(ctx, clientType)
 
-	if err := clientState.Initialize(ctx, k.cdc, clientStore, consensusState); err != nil {
+	clientModule, found := k.router.GetRoute(clientID)
+	if !found {
+		return "", errorsmod.Wrap(types.ErrRouteNotFound, clientID)
+	}
+
+	if err := clientModule.Initialize(ctx, clientID, clientState, consensusState); err != nil {
 		return "", err
 	}
 
-	if status := k.GetClientStatus(ctx, clientState, clientID); status != exported.Active {
+	if status := k.GetClientStatus(ctx, clientID); status != exported.Active {
 		return "", errorsmod.Wrapf(types.ErrClientNotActive, "cannot create client (%s) with status %s", clientID, status)
 	}
 
-	k.Logger(ctx).Info("client created at height", "client-id", clientID, "height", clientState.GetLatestHeight().String())
+	initialHeight := clientModule.LatestHeight(ctx, clientID)
+	k.Logger(ctx).Info("client created at height", "client-id", clientID, "height", initialHeight.String())
 
 	defer telemetry.IncrCounterWithLabels(
 		[]string{"ibc", "client", "create"},
 		1,
-		[]metrics.Label{telemetry.NewLabel(types.LabelClientType, clientState.ClientType())},
+		[]metrics.Label{telemetry.NewLabel(types.LabelClientType, clientType)},
 	)
 
-	emitCreateClientEvent(ctx, clientID, clientState)
+	emitCreateClientEvent(ctx, clientID, clientType, initialHeight)
 
 	return clientID, nil
 }
 
 // UpdateClient updates the consensus state and the state root from a provided header.
 func (k Keeper) UpdateClient(ctx sdk.Context, clientID string, clientMsg exported.ClientMessage) error {
-	clientState, found := k.GetClientState(ctx, clientID)
-	if !found {
-		return errorsmod.Wrapf(types.ErrClientNotFound, "cannot update client with ID %s", clientID)
+	if status := k.GetClientStatus(ctx, clientID); status != exported.Active {
+		return errorsmod.Wrapf(types.ErrClientNotActive, "cannot update client (%s) with status %s", clientID, status)
 	}
 
-	clientStore := k.ClientStore(ctx, clientID)
+	clientType, _, err := types.ParseClientIdentifier(clientID)
+	if err != nil {
+		return errorsmod.Wrapf(types.ErrClientNotFound, "clientID (%s)", clientID)
```

Comment:
error seems misleading to me?
```suggestion
		return errorsmod.Wrapf(err, "unable to parse client identifier %s", clientID)
```